### PR TITLE
New drivers for Brother wordprocessors/typewriters LW-350, LW-450

### PR DESCRIPTION
**Pull request by BartmanAbyss**
split off of [PR#8202](https://github.com/mamedev/mame/pull/8202)
both drivers are still in the same source file because they both access the custom high-level emulated HD63266F.

### CODE REVIEW

File: src/mame/brother/lw350.cpp [hunk 1]
```diff
@@ -0,0 +1,1749 @@
...
+#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+#include "machine/timer.h"
+#include "cpu/z180/z180.h"
+#include "debug/debugcpu.h"
+#include "sound/beep.h"
+#include "video/mc6845.h"
+#include "util/utf8.h"
+#include "imagedev/floppy.h"
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:14]: Same comment as the other drivers – please put these in the conventional order.

File: src/mame/brother/lw350.cpp [hunk 2]
```diff
@@ -0,0 +1,1749 @@
...
+
+// if updating project, c:\msys64\win32env.bat
+// cd \schreibmaschine\mame_src
+// make SUBTARGET=schreibmaschine NO_USE_MIDI=1 NO_USE_PORTAUDIO=1 vs2019
+
+// command line parameters:
+// -log -debug -window -intscalex 2 -intscaley 2 lw350 -resolution 960x256 -flop roms\lw350\Brother_LW-200-300_GW-24-45_Ver1.0_SpreadsheetProgramAndDataStorageDisk.img
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:21]: As already noted, personal usage notes like this don’t belong in the source.

File: src/mame/brother/lw350.cpp [hunk 3]
```diff
@@ -0,0 +1,1749 @@
...
+			fwrite(image.get(), image_length);
+	}
+
+	loaded = false;
+}
+
+// FDC high-level emulation (not timing accurate) based on "Hitachi 8-Bit Microcomputer HD63265 FDC Floppy Disk Controller User's Manual"
+// https://archive.org/details/bitsavers_hitachidatDiskControllerUsersManual2edMar89_3858532
+
+class hd63266f_t : public device_t {
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:170]: Please use names ending in `_device` for `device_t` implementations.

File: src/mame/brother/lw350.cpp [hunk 4]
```diff
@@ -0,0 +1,1749 @@
...
+		logerror("%s: unmapped memory read from %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, mem_mask);
+		return 0;
+	}
+	void illegal_w(offs_t offset, uint8_t data, uint8_t mem_mask = ~0) {
+		logerror("%s: unmapped memory write to %0*X = %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, data, 2, mem_mask);
+	}
+
+	uint8_t rom72000_r(offs_t offset, uint8_t mem_mask = ~0) {
+		return rom[0x02000 + offset] & mem_mask;
+	}
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:1427]: Same comments about log spam with debugger windows open and `mem_mask` in 8-bit memory access handlers apply.

File: src/mame/brother/lw350.cpp [hunk 5]
```diff
@@ -0,0 +1,1749 @@
...
+			return ~0x00; // german
+		default:   return 0x00;
+		}
+	}
+	void io_b8_w(uint8_t data) {
+		io_b8 = data;
+	}
+	void rombank_w(uint8_t data) { // E0
+		membank("rom")->set_entry(data & 0x03);
+	}
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:944]: Please use `required_memory_bank` to look these up so you don’t need to do tag lookups on accesses.

File: src/mame/brother/lw350.cpp [hunk 6]
```diff
@@ -0,0 +1,1749 @@
...
+
+	PORT_START("kbrow.7")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("SM/Layout")             PORT_CODE(KEYCODE_PRTSCR)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("STORNO")                PORT_CODE(KEYCODE_PAUSE)      PORT_CHAR(UCHAR_MAMEKEY(CANCEL))
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_ENTER)      PORT_CHAR(UCHAR_MAMEKEY(ENTER))
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_BACKSPACE)  PORT_CHAR(UCHAR_MAMEKEY(BACKSPACE))
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    //PORT_CODE(KEYCODE_)
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Horz/Vert")             //PORT_CODE(KEYCODE_)
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_SPACE)      PORT_CHAR(UCHAR_MAMEKEY(SPACE))
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD)                                    PORT_CODE(KEYCODE_LSHIFT)     PORT_CHAR(UCHAR_MAMEKEY(LSHIFT))
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:1148]: Same comments about keyboards apply as the other drivers.

File: src/mame/brother/lw350.cpp [hunk 7]
```diff
@@ -0,0 +1,1749 @@
...
+		screen(*this, "screen"),
+		palette(*this, "palette"),
+		floppy(*this, "floppy"),
+		fdc(*this, "fdc"),
+		m_crtc(*this, "hd6445"),
+		vram(*this, "vram"),
+		speaker(*this, "beeper"),
+		io_kbrow(*this, "kbrow.%u", 0),
+		rom(*this, "maincpu")
+	{ }
```
Comment (cuavas) [inline src/mame/brother/lw350.cpp:1396]: Can the two driver state classes use a common base class, or do they really have nothing in common?

**Review by cuavas — COMMENTED** [review]
This driver was the main sticking point last time.  So the things we need to get to the bottom of are:

* Why is it that the µPD765-based implementation doesn’t work?  We‘ve got a lot of systems using variants of the chip, so it would be good if we could get these ones to fit in with the ecosystem.  Maybe it’s something @galibert or @cracyc could help with debugging?
* If the µPD765-based implementation can’t be made to work, can the local HLE implementation of the floppy controller be made to work with MAME’s floppy system?  It would be better if it could be use as much common code as possible.  These local implementations have a habit of sticking around far longer than they should (for an egregious example, look at src/devices/cpu/alto2 – there’s a whole emulator ported to run as a MAME CPU core in there, and it’s been kept on life support across multiple refactorings of the framework).
* If the local HLE implementation *can’t* work with MAME’s floppy system, can it be stripped down to the minimum, i.e. not ape the floppy connector and floppy image device, and just have a single device implementing `device_image_interface` to keep it as lightweight as possible.

**Comment by cuavas** [comment]
@BartmanAbyss I’d recommend looking at the other three drivers first – they’re fairly non-contentious, and the work to get them ready to me merged is pretty straightforward.

File: src/mame/brother/lw350.cpp (hunk 7 above)
Comment (BartmanAbyss) [reply src/mame/brother/lw350.cpp:1396]: currently they only share keyboard and floppy, different video, different memory map, different roms.

Comment (BartmanAbyss) [reply src/mame/brother/lw350.cpp:1396]: although the keyboard should probably be shared between LW-30 and LW-450 rather.

**Comment by BartmanAbyss** [comment]
> This driver was the main sticking point last time. So the things we need to get to the bottom of are:
> 
> * Why is it that the µPD765-based implementation doesn’t work?  We‘ve got a lot of systems using variants of the chip, so it would be good if we could get these ones to fit in with the ecosystem.  Maybe it’s something @galibert or @cracyc could help with debugging?

Dunno, it's been a couple of years. Maybe something with the DEND line not being hooked up to the UPD765?

> * If the µPD765-based implementation can’t be made to work, can the local HLE implementation of the floppy controller be made to work with MAME’s floppy system?  It would be better if it could be use as much common code as possible.  These local implementations have a habit of sticking around far longer than they should (for an egregious example, look at src/devices/cpu/alto2 – there’s a whole emulator ported to run as a MAME CPU core in there, and it’s been kept on life support across multiple refactorings of the framework).

Sure, can you give me some pointers? It should only require a buffer containing the disk image file.
When I last looked at MAME's floppy system, it seems it always wants to convert disk image files to raw bitstreams etc which the HLE emulation doesn't require.

> * If the local HLE implementation _can’t_ work with MAME’s floppy system, can it be stripped down to the minimum, i.e. not ape the floppy connector and floppy image device, and just have a single device implementing `device_image_interface` to keep it as lightweight as possible.

Sure, didn't find a way back then to do it without. Any pointers?

**Comment by ajrhacker** [comment]
DEND almost certainly should be hooked up to or gated onto the FDC's TC line. Systems using the NEC µPD765 family basically need to have some method of generating a pulse on TC, even when they don't do DMA.

**Comment by cuavas** [comment]
> > This driver was the main sticking point last time. So the things we need to get to the bottom of are:
> >
> > * Why is it that the µPD765-based implementation doesn’t work?  We‘ve got a lot of systems using variants of the chip, so it would be good if we could get these ones to fit in with the ecosystem.  Maybe it’s something @galibert or @cracyc could help with debugging?
>
> Dunno, it's been a couple of years. Maybe something with the DEND line not being hooked up to the UPD765?

Well, ideally someone can help get the proper floppy controller working.  Can someone volunteer to help debug what’s going on?  Maybe @cracyc?

> > * If the µPD765-based implementation can’t be made to work, can the local HLE implementation of the floppy controller be made to work with MAME’s floppy system?  It would be better if it could be use as much common code as possible.  These local implementations have a habit of sticking around far longer than they should (for an egregious example, look at src/devices/cpu/alto2 – there’s a whole emulator ported to run as a MAME CPU core in there, and it’s been kept on life support across multiple refactorings of the framework).
>
> Sure, can you give me some pointers? It should only require a buffer containing the disk image file. When I last looked at MAME's floppy system, it seems it always wants to convert disk image files to raw bitstreams etc which the HLE emulation doesn't require.

The floppy system always works in terms of flux, so it’s not going to give you the cooked data.  I guess the high-level simulation isn’t going to work with the floppy system if it needs to work purely with cooked data.

> > * If the local HLE implementation _can’t_ work with MAME’s floppy system, can it be stripped down to the minimum, i.e. not ape the floppy connector and floppy image device, and just have a single device implementing `device_image_interface` to keep it as lightweight as possible.
>
> Sure, didn't find a way back then to do it without. Any pointers?

Well, to make a device that can load media images, all you need to do is implement `device_image_interface`.  The “connector” classes exist to allow the user to select between multiple drive types (`device_slot_interface` provides the glue for user-selectable devices).  You only have your 3.5" drive to deal with, so you don’t need that extra layer.

You should be able to jam all the logic into your high-level HD63265 simulation device.  Make it derive from `device_t` and `device_image_interface`, and put all the logic into it without all the extra layers.

**Comment by cracyc** [comment]
> Dunno, it's been a couple of years. Maybe something with the DEND line not being hooked up to the UPD765?

Presumably dend should be connected to tc.

**Comment by BartmanAbyss** [comment]
Thanks for the pointers. Will have another look at it next week.

**Comment by cracyc** [comment]
Rebased and committed

**Pull request closed without merging**